File: src/main/java/org/sonar/plugins/xquery/checks/DynamicFunctionCheck.java
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks;

import org.sonar.api.rule.RuleKey;
import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.plugins.xquery.parser.XQueryParser;
import org.sonar.plugins.xquery.parser.XQueryTree;
import org.sonar.plugins.xquery.rules.CheckClasses;

/**
 * Checks for usage of dynamic functions.
 * 
 * @since 1.0
 */
@Rule(
    key = DynamicFunctionCheck.RULE_KEY,
    name = "Dynamic Function Usage (Marklogic)",
    description = "Avoid using xdmp:eval() and xdmp:value() where possible. " +
            "Instead use either xdmp:invoke(), xdmp:unpath() " +
            "or if possible assign functions to variables to dynamically evaluate code logic.\n" +
            "Please note that this check is Marklogic specific.",
    priority = Priority.MAJOR)
public class DynamicFunctionCheck extends AbstractCheck {

    public static final String RULE_KEY = "DynamicFunction";
    private static final RuleKey RULE = RuleKey.of(CheckClasses.REPOSITORY_KEY, RULE_KEY);

    @Override
    public void enterExpression(XQueryTree node) {
        super.enterExpression(node);
        if (XQueryParser.FunctionCall == node.getType()) {
            String function = node.getValue("FunctionName.QName");
            if ("xdmp : eval".equals(function) || "xdmp : value".equals(function)) {
                createIssue(RULE, node.getLine());
            }
        }
    }
}

File: src/main/java/org/sonar/plugins/xquery/checks/DynamicFunctionCheck.kt
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks

import org.antlr.v4.runtime.ParserRuleContext
import org.sonar.api.rule.RuleKey
import org.sonar.check.Rule
import org.sonar.plugins.xquery.checks.DynamicFunctionCheck.Companion.RULE_KEY
import org.sonar.plugins.xquery.parser.XQueryParser.FunctionCallContext
import org.sonar.plugins.xquery.parser.getLine
import org.sonar.plugins.xquery.rules.CheckClasses

/**
 * Checks for usage of dynamic functions.
 *
 * @since 1.0
 */
@Rule(key = RULE_KEY)
class DynamicFunctionCheck : AbstractCheck() {

    override fun enterExpression(node: ParserRuleContext) {
        super.enterExpression(node)
        if (node is FunctionCallContext) {
            val function = node.functionName().text
            if ("xdmp:eval" == function || "xdmp:value" == function) {
                createIssue(RULE, node.getLine())
            }
        }
    }

    companion object {
        const val RULE_KEY = "DynamicFunction"
        private val RULE = RuleKey.of(CheckClasses.REPOSITORY_KEY, RULE_KEY)
    }
}